In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

In [2]:
df = pd.read_csv("data_mvreg.csv")
df.head()

Car       Model  Volume  Weight  CO2
0      Toyoty        Aygo    1000     790   99
1  Mitsubishi  Space Star    1200    1160   95
2       Skoda      Citigo    1000     929   95
3        Fiat         500     900     865   90
4        Mini      Cooper    1500    1140  105

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 36 entries, 0 to 35
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Car     36 non-null     object
 1   Model   36 non-null     object
 2   Volume  36 non-null     int64 
 3   Weight  36 non-null     int64 
 4   CO2     36 non-null     int64 
dtypes: int64(3), object(2)
memory usage: 1.5+ KB


In [4]:
df.describe()

Volume       Weight         CO2
count    36.000000    36.000000   36.000000
mean   1611.111111  1292.277778  102.027778
std     388.975047   242.123889    7.454571
min     900.000000   790.000000   90.000000
25%    1475.000000  1117.250000   97.750000
50%    1600.000000  1329.000000   99.000000
75%    2000.000000  1418.250000  105.000000
max    2500.000000  1746.000000  120.000000

In [5]:
df.corr()

Volume    Weight       CO2
Volume  1.000000  0.753537  0.592082
Weight  0.753537  1.000000  0.552150
CO2     0.592082  0.552150  1.000000

## Understanding the correlation between the numeric values:

### There is a positive correlation between volume and weight, when the volume of the engine increases then it's weight increases as well

### There is a positive correlation between volume and CO2 gas, when the volume increases then the corresponding CO2 gad amount increases

#### NOTE: figuring out the meaning of data and correlation and which data is dpenedent or independent:
#### The engine motor is fixed and not going to change (volume will not change neither the weight) but the CO2 gas will change corresponding to enginer specifications.

## Then make a list of the independent values and call this variable X.
## Put the dependent values in a variable called y.

In [9]:
## independent varaibles are motor volume and motor weight
X = df[["Volume","Weight"]]
Y = df[["CO2"]]

In [10]:
X

Volume  Weight
0     1000     790
1     1200    1160
2     1000     929
3      900     865
4     1500    1140
5     1000     929
6     1400    1109
7     1500    1365
8     1500    1112
9     1600    1150
10    1100     980
11    1300     990
12    1000    1112
13    1600    1252
14    1600    1326
15    1600    1330
16    1600    1365
17    2200    1280
18    1600    1119
19    2000    1328
20    1600    1584
21    2000    1428
22    2100    1365
23    1600    1415
24    2000    1415
25    1500    1465
26    2000    1490
27    2000    1725
28    1600    1523
29    2000    1705
30    2100    1605
31    2000    1746
32    1600    1235
33    1600    1390
34    1600    1405
35    2500    1395

In [11]:
Y

CO2
0    99
1    95
2    95
3    90
4   105
5   105
6    90
7    92
8    98
9    99
10   99
11  101
12   99
13   94
14   97
15   97
16   99
17  104
18  104
19  105
20   94
21   99
22   99
23   99
24   99
25  102
26  104
27  114
28  109
29  114
30  115
31  117
32  104
33  108
34  109
35  120

### now creating the regressor:

In [16]:
from sklearn.linear_model import LinearRegression
reg = LinearRegression()
reg.fit(X,Y)

LinearRegression()

### Predicting Results of CO2 for corresponding motor(volume and weight)

In [18]:
#predict the CO2 emission of a car where the weight is 2300kg, and the volume is 1300cm3:
reg.predict([[1300, 2300]]) ## hence the order 1st volume then 2nd is weight.

C:\Users\Osama\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[107.2087328]])

#### The model predicted that a car with 1.3 Litres engine capacity and weight of 2300 kg would produce 107.2 kg of carbon dioxide gas

## Now the coefficients

In [19]:
reg.coef_

array([[0.00780526, 0.00755095]])

#### The result array represents the coefficient values of volume and weight

#### Weight: 0.00755095
#### Volume: 0.00780526

#### These values tell us that if the weight increase by 1kg, the CO2 emission increases by 0.00755095g.

#### And if the engine size (Volume) increases by 1 cm3, the CO2 emission increases by 0.00780526 g.

In [20]:
# We have already predicted that if a car with a 1300cm3 engine weighs 2300kg, the CO2 emission will be approximately 107g.
# What if we increase the weight with 1000kg?

### The equation now is:
#### Assuming CO2 is Y
#### X1 = volume, X2 = weight
### Equation is CO2 = 0.00780526 * volume + 0.00755095 * weight

In [25]:
# to predict new value depending on one of the independent values, we must not change the other value.
0.00780526 * 1300 + 0.00755095 * 3300

35.064972999999995

In [22]:
reg.predict([[1300, (2300+1000)]])

C:\Users\Osama\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[114.75968007]])